clang/lib/CIR/Dialect/IR/CIRTypes.cpp

```diff
@@ -389,6 +389,64 @@ IntType::verify(llvm::function_ref<mlir::InFlightDiagnostic()> emitError,
   return mlir::success();
 }
 
+//===----------------------------------------------------------------------===//
+// FuncType Definitions
+//===----------------------------------------------------------------------===//
+
+mlir::LogicalResult
+FuncType::verify(llvm::function_ref<mlir::InFlightDiagnostic()> emitError,
+                 llvm::ArrayRef<mlir::Type> inputs,
+                 llvm::ArrayRef<mlir::Type> results, bool varArg) {
+  if (results.size() > 1)
+    return emitError() << "functions only supports 0 or 1 results";
+  if (varArg && inputs.empty())
+    return emitError() << "functions must have at least one non-variadic input";
+  return mlir::success();
+}
+
+FuncType FuncType::clone(TypeRange inputs, TypeRange results) const {
+  return get(getContext(), results, inputs, isVarArg());
+}
+
+static mlir::ParseResult
+parseFuncTypeArgs(mlir::AsmParser &p, llvm::SmallVector<mlir::Type> &params,
+                  bool &isVarArg) {
+  isVarArg = false;
+  // `(` `)`
+  if (succeeded(p.parseOptionalRParen()))
+    return mlir::success();
+
+  // type (`,` type)* (`,` `...`)?
+  mlir::Type type;
```

Comment:
Neat parsing!

Reply:
Kudos to [LLVM Dialect](https://github.com/llvm/llvm-project/blob/13e3d4aa5a4e9062bb377207e522cd4f8ed688bd/mlir/lib/Dialect/LLVMIR/IR/LLVMTypes.cpp#L36-L37) :smile: